cookbook/generate_ligand_network.ipynb

```diff
@@ -0,0 +1,357 @@
+    "        max3d=1.0, # Forbid mapping between atoms more than 1.0 Å apart\n",
+    "        element_change=True, # Allow mappings that change an atoms element\n",
+    "        seed='', # Empty SMARTS string causes MCS search to start from scratch\n",
+    "        shift=False, # Keep pre-aligned atom positions for 3D position checks\n",
+    "    ),\n",
+    "    # openfe.setup.PersesAtomMapper(\n",
+    "    #     allow_ring_breaking=True, # Propose mappings that break cyclic systems\n",
+    "    #     preserve_chirality=True, # Forbid mappings that change stereochemistry\n",
+    "    #     use_positions=True, # Use atom positions rather than MCS to choose mappings\n",
+    "    #     coordinate_tolerance=0.25 * unit.angstrom, # Forbid mappings between distant atoms\n",
+    "    # ),\n",
+    "]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "409a3866-a5ac-4295-b91f-3101dd646dc9",
+   "metadata": {},
+   "source": [
+    "## Choose a scoring function"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5998c002-4d39-42c0-b37f-bd0357f27fe6",
+   "metadata": {},
+   "source": [
+    "Planning a `LigandNetwork` requires some way to score each possible network so that the optimal network can be identified. OpenFE scores networks by the quality of the mappings between ligands - transformations that are expected to converge quickly without introducing artifacts are scored highly, and more drastic or risky alchemy is penalised. A scoring function is a function that takes an atom mapping and returns a float between 0.0 and 1.0, with higher numbers representing better maps."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "3bbf3f5b-9e29-467f-aff8-07e8dad7fd4c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# The product of other LOMAP scorers\n",
+    "scorer = openfe.lomap_scorers.default_lomap_score \n",
+    "\n",
+    "# # The default PERSES scoring function\n",
+    "# scorer = openfe.perses_scorers.default_perses_score \n",
+    "\n",
+    "# # Or define your own scoring function\n",
+    "# scorer = lambda m: (\n",
+    "#     openfe.lomap_scorers.default_lomap_score(m) \n",
+    "#     * openfe.perses_scorers.default_perses_scorer(m)\n",
+    "# )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c8c40d49-9657-4691-b6b3-d3c133e215d2",
+   "metadata": {},
+   "source": [
+    "## Plan the network"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "10232f7b-7aac-4773-8c39-15f672a654b5",
+   "metadata": {},
+   "source": [
+    "Finally, combine the ligands, mapper and scorer with a planning function to generate the network. These planning functions usually generate a large number of candidate networks according to some criteria, score them by summing the scores of their constituent mappings, and return the network with the highest score. They differ by the criteria used to construct the candidate networks."
```

Comment:
Hmmm.. I don't have a specific suggestion here, but I think the description isn't quite accurate. Essentially, what's going on here is that more edges means more computational cost, but also better estimates (until the edges selected are really bad). So there's a trade-off of accuracy for computational time. Different network planners use different metrics of how to balance that trade-off.

MST is a strict "minimal compute" solution. It also isn't a very good solution, from what we've been seeing.

Radial is usually based on some external knowledge, and isn't even based on scores (doesn't require them).

We're getting into adding network planners with more complex heuristics now.

Reply:
Ok good to know! I think I can see how I can rewrite this to be more helpful/accurate. I think when I wrote it I was trying to be vague to accommodate future planners - you can totally think of a radial network as being the network that optimises the criteria "every ligand is connected to the central ligand" :sweat_smile: 

But I would like to understand this more fully! So here's some stream of consciousness of how I understand this, and then I'll ask some focused questions in case of tl;dr

-----

My intuition is that an MST and a radial network both have n-1 edges for n nodes - there seems to be an inductive argument that one node has no edges and adding a node must add an edge. Is that incorrect?

So I'm guessing you mean more edges *between two nodes* as well as more edges *in the network* makes things cheaper but less accurate? I'm guessing that an MST allows some edges to have a simpler/better mapping (and so the simulations converge faster?), but a radial network guarantees that any two nodes are only two edges apart. I can see that if all edges have similar precision then the more edges separating two nodes, the less precise the free energy estimate between them, as the error from each edge would compound. This seems to suggest for these examples with very simple transformations a radial graph might be better than an MST?

It seems from documenting the `ProtocolSettings` that the compute time is fixed by default, so faster convergence would mean better precision rather than faster compute? I can see that a mapping that destroys fewer atoms would have a small effect on the number of interactions in the simulation's hybrid states, but this seems negligible for performance. I can also see that extreme transformations just aren't going to work at all and will need some intermediate nodes, and there's a spectrum between "everything works" and "doesn't work at all" where convergence is slow and precision suffers. 

I can see that having more edges *in the network* would improve precision at the cost of performance. Adding an edge is essentially adding a statistical replica, and adding an edge between two distant nodes could improve a lot of errors with relatively little computation. How common are redundant edges in practice? 

I'm guessing your choice probably also depends on which edges you actually care about. If you want to compare a bunch of derivatives to a known binder, a radial network might be perfect. If you're, idk, studying the evolution of a natural product, maybe you want a more linear network. Is it ever worth adding a redundant edge to a connected graph if I don't care about the result of that edge?

Are there other ways that the mapping and network layout affects computational expense?

Is there another sense in which the network layout and mappings affect accuracy, or is accuracy more or less synonymous with precision in this context? Can having the "wrong" edges introduce bias?

So I guess it seems like you usually want something close to a radial network, with all the ligands only a few steps from each other. The exception is when there are, say, two different classes of ligand, and a radial network will converge slowly or not at all. But using an MST for a bunch of simple benzene derivatives is misleading coz actually all the transformations have high scores, and an MST artificially spreads them out. So you might want something like a hub and spoke heuristic, where you group ligands into disconnected radial networks and then try to connect them with like a minimal spanning path.

-----

Stream of consciousness over. Answers to those questions would be very useful but I know that's a lot to write. I'd appreciate any mistakes in the above being corrected. Here are some more focused questions:

1. Does a radial network ever have a different number of edges than an MST?
2. Is the performance and accuracy of a simulation very sensitive to the atom mapping, or is the mapping score more a proxy for "is this transformation going to work at all"?
3. Do you ever, in practice, want redundant edges (ie, a cyclic network), or are the more complex heuristics just about where to put your *n-1* edges?
4. Are there ways that the network layout and atom mapping affect the accuracy of the result directly, or is it just about convergence/precision? Obvs an extremely imprecise result is going to be inaccurate, I just want to confirm that in this context we're talking about how things converge rather than what the underlying model says.

Reply:
I'll tag in @IAlibay to correct me on anything I'm wrong about here.

I'll start with the questions at the bottom:

> Does a radial network ever have a different number of edges than an MST?

Both are the minimum number of edges to span the nodes, so no. At least, not as we define "radial network." There are other tools that start with the kind of radial network we have and then add in more edges (and still call it a "radial network"/"star map").

> Is the performance and accuracy of a simulation very sensitive to the atom mapping, or is the mapping score more a proxy for "is this transformation going to work at all"?

A bit of both? Convergence can definitely depend on the mapping. But also, mapping score itself may not be perfectly correlated with convergence. The scores are just heuristics that people have come up with. A really bad score will probably be a really bad edge, but two scores that are close to each other? May be hard to guess which will actually converge more quickly.

> Do you ever, in practice, want redundant edges (ie, a cyclic network), or are the more complex heuristics just about where to put your n-1 edges?

Yes, in fact most algorithms include cycles. The main reason: once you have a cycle, you know that the sum of the delta-Gs around that cycle should be 0. This can help you identify bad legs, and can be used to get a better maximum likelihood estimate.

> Are there ways that the network layout and atom mapping affect the accuracy of the result directly, or is it just about convergence/precision? Obvs an extremely imprecise result is going to be inaccurate, I just want to confirm that in this context we're talking about how things converge rather than what the underlying model says

Here I definitely want @IAlibay to check me, but I believe it is the case that, in the limit of infinite sampling, any edge should converge to whatever the force field predicts. But yeah, how close to "infinite" your sampling needs to be before you can trust the results. There are pathological cases (e.g., your bound pose for your ligand is completely incorrect) where you might be better off with plain MD.

---

Answering a couple other points:

> This seems to suggest for these examples with very simple transformations a radial graph might be better than an MST?

You argued well for the radial; the counter-argument for MST is that even radial can force you to use edges that only allow bad mappings.

> It seems from documenting the ProtocolSettings that the compute time is fixed by default, so faster convergence would mean better precision rather than faster compute? 

When we talk about faster convergence, it's also "faster" wrt to MD time, not wall time. We'll worry about ns/day inside the protocol, but it is how much MD do you need to do. Two ways this relates to wall time: (1) we have an early-abort, so if your precision is good enough, you can stop early; (2) sometimes you need to continue a simulation for better stats (we haven't made that easy yet, but it is on the way).

> Is it ever worth adding a redundant edge to a connected graph if I don't care about the result of that edge?
> Can having the "wrong" edges introduce bias?

For both of these, keep in mind that, in practice, we're calculating relative free energies and then using maximum likelihood to estimate the absolute free energies. So adding another good edge, that you don't care about, will improve your MLE result. And adding a really bad edge, especially if you have few cycles, can definitely bias you away from correctness.